include/tensor.tcc

```diff
@@ -113,140 +136,154 @@ void* Tensor<T>::data_view(const MemoryLayout mem_layout)
                                             const size_t* dims,
                                             const size_t n_dims,
                                             void* dest_data)
 {
-    if(n_dims > 1) {
+    if (n_dims > 1) {
         T** current = (T**)src_data;
-        for(size_t i = 0; i < dims[0]; i++) {
+        for (size_t i = 0; i < dims[0]; i++) {
           dest_data =
-            this->_copy_nested_to_contiguous(*current, &dims[1],
+            _copy_nested_to_contiguous(*current, &dims[1],
                                              n_dims-1, dest_data);
           current++;
         }
     }
     else {
-        std::memcpy(dest_data, src_data, sizeof(T)*dims[0]);
+        std::memcpy(dest_data, src_data, sizeof(T) * dims[0]);
         return &((T*)dest_data)[dims[0]];
     }
     return dest_data;
 }
 
+// copy from a flat, contiguous memory structure to a provided nested structure
 template <class T>
 void Tensor<T>::_fill_nested_mem_with_data(void* data,
                                            size_t* dims,
                                            size_t n_dims,
                                            size_t& data_position,
                                            void* tensor_data)
 {
-    if(n_dims > 1) {
-        T** current = (T**) data;
-        for(size_t i = 0; i < dims[0]; i++) {
-            this->_fill_nested_mem_with_data(
+    if (n_dims > 1) {
+        T** current = (T**)data;
+        for (size_t i = 0; i < dims[0]; i++, current++) {
+            _fill_nested_mem_with_data(
                     *current, &dims[1], n_dims-1,
                     data_position, tensor_data);
-            current++;
         }
     }
     else {
         T* data_to_copy = &(((T*)tensor_data)[data_position]);
-	    std::memcpy(data, data_to_copy, dims[0]*sizeof(T));
+	    std::memcpy(data, data_to_copy, dims[0] * sizeof(T));
         data_position += dims[0];
     }
-    return;
 }
 
+// Builds nested array structure to point to the provided flat, contiguous
+// memory space.  The space is returned via the data input pointer.
 template <class T>
 T* Tensor<T>::_build_nested_memory(void** data,
                                    size_t* dims,
                                    size_t n_dims,
                                    T* contiguous_mem)
 {
-    if(n_dims>1) {
-        T** new_data = this->_c_mem_views.allocate(dims[0]);
-        (*data) = (void*)new_data;
-        for(size_t i=0; i<dims[0]; i++)
-            contiguous_mem =
-                this->_build_nested_memory((void**)(&new_data[i]),
-                                           &dims[1], n_dims-1,
-                                           contiguous_mem);
+    if (dims == NULL) {
+        throw std::runtime_error("Missing dims in call to "\
+                                 "_build_nested_memory");
+    }
+    if (n_dims > 1) {
+        T** new_data = _c_mem_views.allocate(dims[0]);
+        if (new_data == NULL)
```

Comment:
Same comment as before on non-standard check.  But also, this check seems like it should go inside of ``SharedMemoryList`` near the ``new`` command and not here.

Reply:
This case is different -- we don't know what failure conditions _c_mem_views.allocate() has, so the NULL check is actually just validating the return value. Basic sanity checking on all returned pointers is good defensive practice